JEP 445 support: Unnamed Classes and Instance Main Methods

JEP 445 support for JBang. 

Create  `Main.java`   Java file: 

```java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 21
//JAVAC_OPTIONS --enable-preview -source 21
//JAVA_OPTIONS --enable-preview

void main(){
        System.out.println("Hello, World!");
}
```

Then run `jbang run --verbose Main.java`, and statcktrace as following:

```
[jbang] [0:286] jbang version 0.107.0
[jbang] [0:299] Resolving resource ref: Main.java
[jbang] [0:302] Resolved resource ref as: Main.java (cached as: /Users/linux_china/IdeaProjects/java21-in-action/Main.java)
[jbang] [0:359] Looking for JDK: 21
[jbang] [0:363] Using JDK: 21 (21, current, /Users/linux_china/.jbang/cache/jdks/21)
[jbang] [0:363] No build required. Reusing jar from /Users/linux_china/.jbang/cache/jars/Main.java.8f8f49b93db276e6a59a1b145e26592539e968c045885a1b51ea31d3959ca24d/Main.jar
[jbang] [0:365] Looking for JDK: 21
[jbang] [0:366] Using JDK: 21 (21, current, /Users/linux_china/.jbang/cache/jdks/21)
[jbang] [0:366] Looking for JDK: 21
[jbang] [0:366] Using JDK: 21 (21, current, /Users/linux_china/.jbang/cache/jdks/21)
[jbang] [0:368] [ERROR] no main class deduced, specified nor found in a manifest
dev.jbang.cli.ExitException: no main class deduced, specified nor found in a manifest
	at dev.jbang.source.generators.JarCmdGenerator.generateCommandLineList(JarCmdGenerator.java:168)
	at dev.jbang.source.generators.BaseCmdGenerator.generate(BaseCmdGenerator.java:56)
	at dev.jbang.cli.Run.doCall(Run.java:99)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:145)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:21)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at dev.jbang.cli.JBang$3.handle(JBang.java:148)
	at dev.jbang.cli.JBang$3.handle(JBang.java:143)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at dev.jbang.Main.main(Main.java:14)
[jbang] [0:368] If you believe this a bug in jbang, open an issue at https://github.com/jbangdev/jbang/issues
```

You can run it with `jbang run --verbose -m Main Main.java`, and it works well. 

JEP 445 has been implemented on https://github.com/openjdk/jdk/commit/98b53c06cfffe7b2e21968546b4aa1523f31132e, and next version of JDK 21 will include this feature.

--- a/itests/karate.java
+++ b/itests/karate.java
@@ -1,5 +1,5 @@
 ///usr/bin/env jbang "$0" "$@" ; exit $?
-//REPOS jitpack
+//REPOS jitpack,mavencentral
 // Using jitpack to get latest dev version of karate to be able
 // to use the just added fork() methods to test clis.
 // see https://github.com/intuit/karate/issues/1191
@@ -7,6 +7,8 @@
 //DEPS com.github.intuit.karate:karate-netty:e2882c4
 ////DEPS com.intuit.karate:karate-netty:2.0.0
 ////DEPS com.github.maxandersen.karate:karate-netty:19e06766
+//DEPS org.openjdk.nashorn:nashorn-core:15.4
+
 
 class karate {
 

--- a/itests/karate.java
+++ b/itests/karate.java
@@ -1,5 +1,5 @@
 ///usr/bin/env jbang "$0" "$@" ; exit $?
-//REPOS jitpack,mavencentral
+//REPOS jitpack
 // Using jitpack to get latest dev version of karate to be able
 // to use the just added fork() methods to test clis.
 // see https://github.com/intuit/karate/issues/1191
@@ -7,7 +7,6 @@
 //DEPS com.github.intuit.karate:karate-netty:e2882c4
 ////DEPS com.intuit.karate:karate-netty:2.0.0
 ////DEPS com.github.maxandersen.karate:karate-netty:19e06766
-//DEPS org.openjdk.nashorn:nashorn-core:15.4
 
 
 class karate {

--- a/itests/nakedmain.feature
+++ b/itests/nakedmain.feature
@@ -1,0 +1,20 @@
+Feature: jep445
+
+Background:
+// as long as java 21 is not downloadable
+// we keep this abort in place - it should be removed
+// until then test by:
+// JBANG_JDK_VENDOR=openjdk ./karate.java jep445.feature
+* if (java.lang.System.getenv('JBANG_JDK_VENDOR') != "openjdk") karate.abort()
+
+Scenario: naked main
+When command('jbang nakedmain/noclass.java')
+Then match out == "Hello!\n"
+
+Scenario: naked main
+When command('jbang nakedmain/noclasswithargs.java hey')
+Then match out == "Hello hey\n"
+
+Scenario: naked main
+When command('jbang nakedmain/classinstance.java yo')
+Then match out == "Hello yo\n"

--- a/itests/nakedmain/classinstance.java
+++ b/itests/nakedmain/classinstance.java
@@ -1,0 +1,8 @@
+//JAVA 21+
+//PREVIEW
+class classinstance {
+    void main(String... args) {
+
+        System.out.println("Hello " + args[0]);
+    }
+}

--- a/itests/nakedmain/noclass.java
+++ b/itests/nakedmain/noclass.java
@@ -1,0 +1,6 @@
+//JAVA 21+
+//PREVIEW
+
+void main() {
+    System.out.println("Hello!");
+}

--- a/itests/nakedmain/noclasswithargs.java
+++ b/itests/nakedmain/noclasswithargs.java
@@ -1,0 +1,6 @@
+//JAVA 21+
+//PREVIEW
+
+void main(String... args) {
+    System.out.println("Hello " + args[0]);
+}

--- a/src/main/java/dev/jbang/source/buildsteps/CompileBuildStep.java
+++ b/src/main/java/dev/jbang/source/buildsteps/CompileBuildStep.java
@@ -281,6 +281,8 @@ public abstract class CompileBuildStep implements Builder<Project> {
 	protected abstract String getMainExtension();
 
 	protected Predicate<ClassInfo> getMainFinder() {
-		return pubClass -> pubClass.method("main", STRINGARRAYTYPE) != null;
+		return pubClass -> (pubClass.method("main", STRINGARRAYTYPE) != null
+				||
+				pubClass.method("main") != null);
 	}
 }

--- a/src/main/java/dev/jbang/util/Util.java
+++ b/src/main/java/dev/jbang/util/Util.java
@@ -80,7 +80,7 @@ public class Util {
 	private static final String JBANG_DOWNLOAD_SOURCES = "JBANG_DOWNLOAD_SOURCES";
 
 	public static final Pattern patternMainMethod = Pattern.compile(
-			"^.*(public\\s+static|static\\s+public)\\s+void\\s+main\\s*\\(.*",
+			"^.*(public\\s+static|static\\s+public)\\s+void\\s+main\\s*\\(.*|void\\s+main\\s*\\(\\)",
 			Pattern.MULTILINE);
 
 	public static final Pattern mainClassMethod = Pattern.compile(

--- a/src/main/scripts/choco/jbang.nuspec
+++ b/src/main/scripts/choco/jbang.nuspec
@@ -1,5 +1,4 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<!-- Do not remove this test for UTF-8: if “Ω” doesn’t appear as greek uppercase omega letter enclosed in quotation marks, you should use an editor that supports UTF-8, not this one. -->
 <package xmlns="http://schemas.microsoft.com/packaging/2015/06/nuspec.xsd">
   <metadata>
     <id>jbang</id>

--- a/src/test/java/dev/jbang/cli/TestJdk.java
+++ b/src/test/java/dev/jbang/cli/TestJdk.java
@@ -14,7 +14,6 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Arrays;
 import java.util.function.BiConsumer;
-import java.util.function.BiFunction;
 import java.util.function.Function;
 
 import org.junit.jupiter.api.BeforeEach;
@@ -442,7 +441,6 @@ class TestJdk extends BaseTest {
 		}
 	}
 
-
 	private void initMockJdkDirRuntime(Path jdkPath, String version) {
 		initMockJdkDir(jdkPath, version, "JAVA_RUNTIME_VERSION");
 	}
@@ -450,6 +448,7 @@ class TestJdk extends BaseTest {
 	private void initMockJdkDir(Path jdkPath, String version) {
 		initMockJdkDir(jdkPath, version, "JAVA_VERSION");
 	}
+
 	private void initMockJdkDir(Path jdkPath, String version, String key) {
 		Util.mkdirs(jdkPath);
 		String rawJavaVersion = key + "=\"" + version + "\"";

--- a/src/test/java/dev/jbang/util/TestUtil.java
+++ b/src/test/java/dev/jbang/util/TestUtil.java
@@ -70,6 +70,21 @@ public class TestUtil extends BaseTest {
 
 		content = "public static\nvoid\nmain(String\nargs[]) {";
 		assertTrue(Util.hasMainMethod(content));
+
+		// the Java 21 / JEP-445
+
+		content = "class HelloWorld {\n" +
+				"\tvoid main() {\n" +
+				"\t\tSystem.out.println(\"Hello world from an instance main method o/\");\n" +
+				"\t}\n" +
+				"}";
+		assertTrue(Util.hasMainMethod(content));
+
+		content = "void main() {\n" +
+				"\t\tSystem.out.println(\"Hello world from an unnamed main method o/\");\n" +
+				"\t}\n";
+		assertTrue(Util.hasMainMethod(content));
+
 	}
 
 	@Test

